sanic/http.py

```diff
@@ -0,0 +1,406 @@
+    ]
+
+    def __init__(self, protocol):
+        self._send = protocol.send
+        self._receive_more = protocol.receive_more
+        self.recv_buffer = protocol.recv_buffer
+        self.protocol = protocol
+        self.expecting_continue = False
+        self.stage = Stage.IDLE
+        self.request_body = None
+        self.request_bytes = None
+        self.request_max_size = protocol.request_max_size
+        self.keep_alive = True
+        self.head_only = None
+        self.request = None
+        self.response = None
+        self.exception = None
+        self.url = None
+
+    def __bool__(self):
+        """Test if request handling is in progress"""
+        return self.stage in (Stage.HANDLER, Stage.RESPONSE)
+
+    async def http1(self):
+        """HTTP 1.1 connection handler"""
+        while True:  # As long as connection stays keep-alive
+            try:
+                # Receive and handle a request
+                self.stage = Stage.REQUEST
+                self.response_func = self.http1_response_header
+                await self.http1_request_header()
+                await self.protocol.request_handler(self.request)
+                # Handler finished, response should've been sent
+                if self.stage is Stage.HANDLER:
+                    raise ServerError("Handler produced no response")
+                if self.stage is Stage.RESPONSE:
+                    await self.response.send(end_stream=True)
+            except CancelledError:
+                # Write an appropriate response before exiting
+                e = self.exception or ServiceUnavailable(f"Cancelled")
+                self.exception = None
+                self.keep_alive = False
+                await self.error_response(e)
+            except Exception as e:
+                # Write an error response
+                await self.error_response(e)
+            # Try to consume any remaining request body
+            if self.request_body:
+                if self.response and 200 <= self.response.status < 300:
+                    logger.error(f"{self.request} body not consumed.")
+                try:
+                    async for _ in self:
+                        pass
+                except PayloadTooLarge:
+                    self.keep_alive = False
+            # Exit and disconnect if no more requests can be taken
+            if self.stage is not Stage.IDLE or not self.keep_alive:
+                break
+            # Wait for next request
+            if not self.recv_buffer:
+                await self._receive_more()
+
+    async def http1_request_header(self):
+        """Receive and parse request header into self.request."""
+        # Receive until full header is in buffer
+        buf = self.recv_buffer
+        pos = 0
+        while True:
+            pos = buf.find(b"\r\n\r\n", pos)
+            if pos != -1:
+                break
+            pos = max(0, len(buf) - 3)
+            if pos >= self.request_max_size:
```

Comment:
Perhaps this should be a separate config. A max size for just the headers. [Nginx defaults to 1k](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_header_buffer_size).

Reply:
I've also considered separate setting for receive buffer size, as there is for websockets. However, for simplicity and ease of config, the number of options really should be kept to bare minimum, i.e. what is actually required in practical applications. Apache limits them to 8 kB and that doesn't seem to bother anyone. Already implemented in this branch: send buffer size (in asyncio protocol) is hardcoded 64 KiB.

Reply:
I am fine also hard coding it. But tying it to the `request_max_size` could be problematic if the dev sets that extremely high to handle large uploads, etc. I say let's just pick something reasonable.